### ISSUE TITLE
Input in chromuim  by  text-input-v3, the pop windows will  blink one time.

### ISSUE THREAD

**Issue by knm100** [labwc/labwc#2170]
The fcitx5 pop window will first blink one time in the upper left corner before returning to its correct position
if comment out `--wayland-text-input-version=3`  in chromium-flags.conf,  it will work fine.
labwc rev:   https://github.com/tokyo4j/labwc#branch=app-icon + text-input-v1 pacth

![swappy-20240922-095010](https://github.com/user-attachments/assets/0365d4c8-2401-47dd-bd0e-8f59af86ae63)

https://github.com/user-attachments/assets/9e8081b3-d59e-4533-9943-1f4ede64ef5c

**Comment by tokyo4j** [comment]
Perhaps https://issues.chromium.org/issues/362191728 fixed this issue. I don't know if it's included in the latest release though.

**Comment by knm100** [comment]
Thanks,   Let‘s wait patiently  for two weeks,  it should fix the problem  in the next chromium stable version.
In my case, linking to  `https://issues.chromium.org` always return `ERR_CONNECTION_REFUSED`

**Comment by knm100** [comment]
the pop windows still blink one time in chromium 130 under labwc,  but not under cosmic-comp

**Comment by tokyo4j** [comment]
> the pop windows still blink one time in chromium 130 under labwc, but not under cosmic-comp

This is because Smithay, the base component of cosmic-comp, doesn't strictly follow the protocol.

The cursor position set by clients with `text_input_v3::set_cursor_rectangle` requests should be double-buffered and applied with `text_input_v3::commit` requests, but Smithay seems to apply it immediately on `text_input_v3::set_cursor_rectangle` requests. Also, the protocol says the pending cursor position should be cleared on `text_input_v3::commit` requests, but remembers the old cursor position.
 
> Edited:
> Under the kwin, the pop windows still blink one time in chromium130. So, chromium should not fix this issue.

Yeah I tested Chromium 130 and confirmed the blink is not fixed. I wonder what problem the Choromium dev fixed in the MR I quoted. When I have some time, I'll investigate it ask them why the issue is still there.

**Comment by tokyo4j** [comment]
I posted a comment in the issue (https://issues.chromium.org/issues/362191728):

<details><summary>comment</summary>

Hello, I'm a maintainer of labwc, a wayland compositor used by Raspberry Pi OS, and have added text-input-v3 support.

For some reason, the popup is still momentarily shown at the top-left corner of the monitor on labwc when I start typing.

Output of `WAYLAND_DEBUG=client chromium 2>&1 | grep text_input` shows that this is because `set_cursor_rectangle` request is sent only after `done` event, but not after `enter` event.

Followings are the outputs when keyboard focus is moved to chromium on each of labwc and kwin:

labwc:
```
[3486173.258] zwp_text_input_v3#60.enter(wl_surface#56)
[3486173.597]  -> zwp_text_input_v3#60.enable()
[3486173.604]  -> zwp_text_input_v3#60.commit()
```

kwin:
```
[3551722.854] zwp_text_input_v3#62.enter(wl_surface#45)
[3551723.343]  -> zwp_text_input_v3#62.enable()
[3551723.351]  -> zwp_text_input_v3#62.commit()
[3551729.646] zwp_text_input_v3#62.done(11)
[3551729.651]  -> zwp_text_input_v3#62.set_content_type(0, 5)
[3551729.654]  -> zwp_text_input_v3#62.set_cursor_rectangle(174, 63, 1, 20)
[3551729.658]  -> zwp_text_input_v3#62.set_surrounding_text("", 0, 0)
[3551729.661]  -> zwp_text_input_v3#62.commit()
[3551730.069] zwp_text_input_v3#62.done(12)
```

As you can see, kwin sends `done` event some milliseconds after `enter` event, then chromium sends `set_cursor_rectangle`, which prevents flickering popup. But labwc doesn't send `done` event, so `set_cursor_rectangle` request is sent only after anything is typed in chromium, which causes flickering popup.

Environment:
- labwc 0.8.2
- kwin 6.2.4
- fcitx5 5.1.11
- fcitx5-mozc 2.26.4632.102.g4d2e3bd-2
- Chromium 131.0.6778.139 Arch Linux

</details>

**Comment by Consolatis** [comment]
> This is because Smithay, the base component of cosmic-comp, doesn't strictly follow the protocol.
> 
> The cursor position set by clients with `text_input_v3::set_cursor_rectangle` requests should be double-buffered and applied with `text_input_v3::commit` requests, but Smithay seems to apply it immediately on `text_input_v3::set_cursor_rectangle` requests. Also, the protocol says the pending cursor position should be cleared on `text_input_v3::commit` requests, but remembers the old cursor position.

JFYI @Drakulix

**Comment by Drakulix** [comment]
> > This is because Smithay, the base component of cosmic-comp, doesn't strictly follow the protocol.
> > The cursor position set by clients with `text_input_v3::set_cursor_rectangle` requests should be double-buffered and applied with `text_input_v3::commit` requests, but Smithay seems to apply it immediately on `text_input_v3::set_cursor_rectangle` requests. Also, the protocol says the pending cursor position should be cleared on `text_input_v3::commit` requests, but remembers the old cursor position.
> 
> JFYI @Drakulix

Thanks, tracked here https://github.com/Smithay/smithay/issues/1614

**Comment by kchibisov** [comment]
>Also, the protocol says the pending cursor position should be cleared on text_input_v3::commit requests, but smithay remembers the old cursor position.

Could you provide a reference for this? This is not mentioned in the protocol anywhere. The position is valid until the next committed `enabled` or `disabled`, not until the next `commit`.

https://wayland.app/protocols/text-input-unstable-v3#zwp_text_input_v3:request:set_cursor_rectangle

**Comment by tokyo4j** [comment]
> Could you provide a reference for this? This is not mentioned in the protocol anywhere. The position is valid until the next committed enabled or disabled, not until the next commit.

You're right. I misunderstood the protocol. But still, from my observation, Smithay seems to remember the cursor rectangle after `disable` and `enable`.

**Comment by kchibisov** [comment]
>You're right. I misunderstood the protocol. But still, from my observation, Smithay seems to remember the cursor rectangle after disable and enable.

that's _kind of_ correct.

**Comment by AbdAlRahmanGad** [comment]
Hi @tokyo4j 

We've been investigating this issue in Chromium and we think that the fix might be needed from the compositor side, could you please take a look at this comment by Orko: https://issues.chromium.org/u/1/issues/384531043#comment6

Extra context: an issue filed in `sway` and there is a open PR sent: https://github.com/swaywm/sway/issues/8884

**Comment by tokyo4j** [comment]
> Hi @tokyo4j 
> 
> We've been investigating this issue in Chromium and we think that the fix might be needed from the compositor side, could you please take a look at this comment by Orko: https://issues.chromium.org/u/1/issues/384531043#comment6
> 
> Extra context: an issue filed in `sway` and there is a open PR sent: https://github.com/swaywm/sway/issues/8884

Oh sorry, I didn't notice the comment by Orko. I'll address it this weekend.

**Comment by AbdAlRahmanGad** [comment]
Thanks, no problem at all. 
Please, take you time.
